File: telegram/src/main/java/by/mrrockka/service/BountyTelegramService.java
package by.mrrockka.service;

import by.mrrockka.domain.MessageMetadata;
import by.mrrockka.domain.Person;
import by.mrrockka.domain.TelegramPerson;
import by.mrrockka.domain.bounty.Bounty;
import by.mrrockka.domain.collection.PersonEntries;
import by.mrrockka.domain.game.BountyGame;
import by.mrrockka.parser.BountyMessageParser;
import by.mrrockka.service.exception.ChatGameNotFoundException;
import by.mrrockka.service.exception.EntriesForPersonNotFoundException;
import by.mrrockka.service.game.GameTelegramFacadeService;
import by.mrrockka.validation.bounty.BountyValidator;
import by.mrrockka.validation.mentions.PersonMentionsValidator;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.telegram.telegrambots.meta.api.methods.botapimethods.BotApiMethodMessage;
import org.telegram.telegrambots.meta.api.methods.send.SendMessage;

import java.util.List;

@Service
@RequiredArgsConstructor
public class BountyTelegramService {

  private final BountyService bountyService;
  private final BountyMessageParser bountyMessageParser;
  private final GameTelegramFacadeService gameTelegramFacadeService;
  private final PersonMentionsValidator personMentionsValidator;
  private final BountyValidator bountyValidator;

  public BotApiMethodMessage storeBounty(final MessageMetadata messageMetadata) {
    personMentionsValidator.validateMessageMentions(messageMetadata, 2);
    final var fromAndTo = bountyMessageParser.parse(messageMetadata);
    final var telegramGame = gameTelegramFacadeService
      .getGameByMessageMetadata(messageMetadata)
      .orElseThrow(ChatGameNotFoundException::new);
    final var game = telegramGame.game().asType(BountyGame.class);

    bountyValidator.validate(game, fromAndTo);

    final var gamePersons = game.getEntries().stream()
      .map(PersonEntries::person)
      .toList();

    final var bounty = Bounty.builder()
      .to(findByNickname(fromAndTo.getValue(), gamePersons))
      .from(findByNickname(fromAndTo.getKey(), gamePersons))
      .amount(game.asType(BountyGame.class).getBountyAmount())
      .build();

    bountyService.storeBounty(game.getId(), bounty, messageMetadata.getCreatedAt());

    return SendMessage.builder()
      .chatId(messageMetadata.getChatId())
      .text("Bounty amount %s from %s stored for %s"
              .formatted(game.getBountyAmount(), fromAndTo.getKey().getNickname(), fromAndTo.getValue().getNickname()))
      .replyToMessageId(telegramGame.messageMetadata().getId())
      .build();
  }

  private Person findByNickname(final TelegramPerson telegramPerson, final List<Person> persons) {
    return persons.stream()
      .filter(person -> person.getNickname().equals(telegramPerson.getNickname()))
      .findFirst()
      .orElseThrow(EntriesForPersonNotFoundException::new);
  }
}

File: telegram/src/main/java/by/mrrockka/service/BountyTelegramService.kt
package by.mrrockka.service

import by.mrrockka.domain.Bounty
import by.mrrockka.domain.BountyTournamentGame
import by.mrrockka.domain.MessageMetadata
import by.mrrockka.domain.takenToGiven
import by.mrrockka.parser.BountyMessageParser
import by.mrrockka.repo.BountyRepo
import org.springframework.stereotype.Service

@Service
class BountyTelegramService(
        private val bountyRepo: BountyRepo,
        private val bountyMessageParser: BountyMessageParser,
        private val gameService: GameTelegramService,
) {

    fun store(metadata: MessageMetadata): Bounty {
        metadata.checkMentions()
        val (from, to) = bountyMessageParser.parse(metadata)
        val telegramGame = gameService.findGame(metadata)
        check(telegramGame.game is BountyTournamentGame) { "Bounties can be submitted only for Bounty tournament" }

        validate(telegramGame.game, from, to)

        val players = telegramGame.game.players.associateBy { it.person.nickname }
        val fromPlayer = players[from]!!
        val toPlayer = players[to]!!

        val bounty = Bounty(from = fromPlayer.person, to = toPlayer.person, amount = telegramGame.game.bounty)
        bountyRepo.store(telegramGame.game.id, bounty, metadata.createdAt)

        return bounty
    }

    private fun validate(game: BountyTournamentGame, from: String, to: String) {
        check(from != to) { "You can't kick yourself off for bounty" }

        val fromPlayer = game.players.find { it.person.nickname == from }
                ?: error("@$from person hadn't enter game")
        val toPlayer = game.players.find { it.person.nickname == to }
                ?: error("@$to person hadn't enter game")
        val fromEntries = fromPlayer.entries
        val toEntries = toPlayer.entries
        val (_, fromGiven) = fromPlayer.takenToGiven()
        val (_, toGiven) = toPlayer.takenToGiven()

        check(fromEntries.size - fromGiven.size > 0) { "@$from knocked off from the game" }
        check(toEntries.size - toGiven.size > 0) { "@$to knocked off from the game" }
    }
}
